### PR TITLE
ref(tests): Introduce React-Testing-Library

### PR DESCRIPTION
**Pull request by priscilawebdev**
Reopening the PR https://github.com/getsentry/sentry/pull/18039 here 

**What:** Introducing React Testing Library

**Why:** We would like to take RTL out for a spin as it seems to bring several advantages over Enzyme, for instance:

- Rather than tests focusing on the implementation (Enzyme), tests are focused on user behavior (RTL) 
- Performance gain and ease of using hooks
- Fewer or no issues when updating React
- No shallow rendering - (more context https://kentcdodds.com/blog/why-i-never-use-shallow-rendering)

**Note:** We will be evaluating the performance of the issue stream tests before any final decision.

### CODE REVIEW

File: tests/js/spec/views/issueList/tagFilter.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @jest-environment jest-environment-jsdom-fifteen
```
Comment (priscilawebdev) [inline tests/js/spec/views/issueList/tagFilter.spec.tsx:2]: This is needed as we are using a Jest version lower than 25. Please see: https://testing-library.com/docs/react-testing-library/setup/#jest-24-or-lower-and-defaults

File: jest.config.js [hunk 2]
```diff
@@ -53,7 +53,10 @@ module.exports = {
     '<rootDir>/tests/js/setup.js',
     'jest-canvas-mock',
   ],
-  setupFilesAfterEnv: ['<rootDir>/tests/js/setupFramework.ts'],
+  setupFilesAfterEnv: [
+    '<rootDir>/tests/js/setupFramework.ts',
+    '@testing-library/jest-dom/extend-expect',
```
Comment (priscilawebdev) [inline jest.config.js:58]: more context: https://github.com/testing-library/jest-dom#usage

File: tests/js/spec/components/avatar/avatarList.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      // @ts-expect-error
+      TestStubs.User({id: '2', name: 'BC'}),
+    ];
+
+    const {container, queryByTestId, getByText} = renderComponent(users);
+    expect(getByText('A')).toBeTruthy();
+    expect(getByText('B')).toBeTruthy();
+    expect(queryByTestId('avatarList-collapsedusers')).toBeNull();
+
+    expect({html: () => container.innerHTML}).toSnapshot();
```
Comment (priscilawebdev) [inline tests/js/spec/components/avatar/avatarList.spec.tsx:23]: This was needed because our visual snapshot package currently only accepts the `html()` enzyme function.

https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/html.html

https://github.com/getsentry/jest-visual-snapshot/blob/main/packages/jest/toSnapshot.js

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 1 above)
Comment (billyvg) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:2]: We're using jest 26

File: static/app/views/issueList/tagFilter.tsx [hunk 4]
```diff
@@ -167,24 +172,38 @@ class IssueListTagFilter extends React.Component<Props, State> {
...
+                label: string;
+                value: string;
+              }>) => {
+                return (
+                  <components.Option label={label} {...optionProps}>
+                    <span data-testid={`select-control-option-${label}`}> {label}</span>
+                  </components.Option>
+                );
+              },
+            }}
```
Comment (billyvg) [inline static/app/views/issueList/tagFilter.tsx:206]: Is this change needed?

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 1 above)
Comment (priscilawebdev) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:2]: @billyvg  I meant here https://github.com/getsentry/jest-visual-snapshot/blob/b97665bc92fa2313ef79196615f9e621518eade4/packages/jest-environment/package.json#L22

Comment (priscilawebdev) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:2]: I will double-check it, but I believe this was the cause of an issue

File: static/app/views/issueList/tagFilter.tsx (hunk 4 above)
Comment (priscilawebdev) [reply static/app/views/issueList/tagFilter.tsx:206]: We can override it :)

I will update the PR with this change

https://testing-library.com/docs/queries/bytestid/#overriding-data-testid

Comment (billyvg) [reply static/app/views/issueList/tagFilter.tsx:206]: @priscilawebdev you could see what's more popular and we can go with that convention

Comment (priscilawebdev) [reply static/app/views/issueList/tagFilter.tsx:206]: Testing Library recommends adopting the attribute `data-testid` where possible. should we follow their recommendation?

Comment (priscilawebdev) [reply static/app/views/issueList/tagFilter.tsx:206]: > following the precedent set by React Native Web which uses a testID prop to emit a data-testid attribute on the element, and we recommend you adopt that attribute where possible

Comment (billyvg) [reply static/app/views/issueList/tagFilter.tsx:206]: I think what's more important than what they recommend is what the community has adopted. (e.g. consider what happens in a few years when we replace RTL :P) - https://github.com/search?q=data-testid&type=code vs https://github.com/search?q=data-test-id&type=code

Comment (priscilawebdev) [reply static/app/views/issueList/tagFilter.tsx:206]: hmm .. I don't see many problems with this, as I believe that replacing this attribute would not be complicated, but for now, as our codebase already uses "data-test-id", I would prefer to keep it

Comment (markstory) [reply static/app/views/issueList/tagFilter.tsx:206]: I don't think there is much value for us in churning `data-test-id` at this point. Reconfiguring RTL to fit with what we have already would be the path I take. Getting folks to change over tests is going to be enough of a chore that we don't need to compound it with changing data-test-id at the same time. If in the future we need to change we can address that problem separately.

File: tests/js/spec/views/issueList/tagFilter.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      />,
+      routerContext
+    );
+
+    // changes dropdown input value
+    const input = getByLabelText(tag.key);
+    fireEvent.change(input, {target: {value: 'foo'}});
+
+    // waits for the loading indicator to disappear
+    const loadingIndicator = getByText('Loading...');
```
Comment (markstory) [inline tests/js/spec/views/issueList/tagFilter.spec.tsx:64]: Should this be using an ellipsis or just `Loading` instead?

File: tests/js/spec/components/avatar/avatarList.spec.tsx (hunk 3 above)
Comment (billyvg) [reply tests/js/spec/components/avatar/avatarList.spec.tsx:23]: I've updated the package to support this (although I haven't tested it) https://github.com/getsentry/jest-visual-snapshot/blob/main/packages/jest/toSnapshot.js#L6

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 1 above)
Comment (billyvg) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:2]: I updated the package to now use jest-environment-jsdom v26

File: tests/js/spec/components/avatar/avatarList.spec.tsx (hunk 3 above)
Comment (priscilawebdev) [reply tests/js/spec/components/avatar/avatarList.spec.tsx:23]: Nice! Thank you 🙏

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 5 above)
Comment (priscilawebdev) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:64]: it should be using ellipsis to match the string, otherwise it won't work

File: tests/js/spec/components/avatar/avatarList.spec.tsx (hunk 3 above)
Comment (priscilawebdev) [reply tests/js/spec/components/avatar/avatarList.spec.tsx]: it works 😉

File: static/app/views/issueList/tagFilter.tsx [hunk 6]
```diff
@@ -167,24 +172,38 @@ class IssueListTagFilter extends React.Component<Props, State> {
...
+              Option: ({
+                label,
+                ...optionProps
+              }: OptionProps<{
+                label: string;
+                value: string;
+              }>) => {
+                return (
+                  <components.Option label={label} {...optionProps}>
+                    <span data-test-id={`select-control-option-${label}`}>{label}</span>
```
Comment (EvanPurkhiser) [inline static/app/views/issueList/tagFilter.tsx:202]: Seems like this should actual be done at our select component level instead of a one off here, no?

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 5 above)
Comment (EvanPurkhiser) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:60]: This is three dots, where as above it's `Loading\u2026` no?

Comment (priscilawebdev) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:60]: `Loading…` does not match `Loading...`, but I will double-check it

Comment (EvanPurkhiser) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:60]: I wonder if it does something like https://www.npmjs.com/package/normalize-diacritics to keep the text ANSI, 🤔 strange

File: static/app/views/issueList/tagFilter.tsx (hunk 6 above)
Comment (priscilawebdev) [reply static/app/views/issueList/tagFilter.tsx:202]: I updated the code with another way to achieve this, maybe it looks better now

File: tests/js/spec/views/issueList/tagFilter.spec.tsx (hunk 5 above)
Comment (priscilawebdev) [reply tests/js/spec/views/issueList/tagFilter.spec.tsx:60]: this was happening because react-select was rendering its default load message, which doesn't use Unicode ellipses. I updated the code and now it works, but a way to ignore the ellipses and just check the load string would be:

`const loadingIndicator = getByText(content => content.startsWith('Loading'));`

**Review by evanpurkhiser — APPROVED** [review]
Looks good to me
